runtime/runtime_impl_jetty12/src/main/java/com/google/apphosting/runtime/jetty/http/JettyHttpHandler.java

```diff
@@ -108,8 +110,25 @@ public boolean handle(Request request, Response response, Callback callback) thr
     ApiProxy.Environment currentEnvironment = ApiProxy.getCurrentEnvironment();
     request.setAttribute(AppEngineConstants.ENVIRONMENT_ATTR, currentEnvironment);
 
+    // Only run code to finish request with the RequestManager once the stream is complete.
+    request.addHttpStreamWrapper(
+        httpStream ->
+            new HttpStream.Wrapper(httpStream) {
+              @Override
+              public void succeeded() {
+                super.succeeded();
+                finishRequest(currentEnvironment, requestToken, genericResponse, context);
+              }
+
+              @Override
+              public void failed(Throwable x) {
+                super.failed(x);
+                finishRequest(currentEnvironment, requestToken, genericResponse, context);
+              }
+            });
+
     try {
-      handled = dispatchRequest(requestToken, genericRequest, genericResponse, callback);
+      handled = dispatchRequest(requestToken, genericRequest, genericResponse);
```

Comment:
technically, if handled is false, then you should disable the completion listener.  It can't be removed, but it can be atomically turned into a noop.    

For GAE, it is not really important, as the request will a) never return false here; b) never be given to another context if it does.... so perhaps just a comment saying that if handled would ever be false, then the listener should be disabled... or maybe just an `assert handled == true`?

Reply:
The other way of doing this would be to also protect the finishedRequest call from being called twice if both succeeded and failed were ever called.   You could have the requestToken set in an AtomicReference and do a getAndSet to null from succeeded, failed and handled==false.

Or just comment... as it should never happen.

Reply:
> if handled is false, then you should disable the completion listener. It can't be removed, but it can be atomically turned into a noop.

I don't see why we'd ever want to turn the completion listener into a noop, we have already called `requestManager.startRequest` so there needs to be a matching `requestManager.finishRequest`.